pkg/controller/core/reconciler.go

```diff
@@ -0,0 +1,457 @@
+		}
+
+		// We've successfully processed the delete, so there's no further
+		// reconciliation to do. There's a good chance our claim no longer
+		// exists, but we try update its status just in case it sticks around,
+		// for example due to additional finalizers.
+		claim.SetConditions(v1alpha1.Deleting(), v1alpha1.ReconcileSuccess())
+		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), "cannot update resource claim status")
+	}
+
+	if !meta.WasCreated(resource) {
+		class := &v1alpha1.ResourceClass{}
+		if err := r.client.Get(ctx, meta.NamespacedNameOf(claim.GetClassReference()), class); err != nil {
+			// If we didn't hit this error last time we'll be requeued
+			// implicitly due to the status update. Otherwise we want to retry
+			// after a brief wait, in case this was a transient error or the
+			// class is (re)created.
+			claim.SetConditions(v1alpha1.Creating(), v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: AShortWait}, errors.Wrap(r.client.Status().Update(ctx, claim), "cannot update resource claim status")
+		}
+
+		if err := r.resource.Configure(ctx, claim, class, resource); err != nil {
+			// If we didn't hit this error last time we'll be requeued
+			// implicitly due to the status update. Otherwise we want to retry
+			// after a brief wait, in case this was a transient error or some
+			// issue with the resource class was resolved.
+			claim.SetConditions(v1alpha1.Creating(), v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: AShortWait}, errors.Wrap(r.client.Status().Update(ctx, claim), "cannot update resource claim status")
+		}
+
+		if err := r.resource.Create(ctx, claim, class, resource); err != nil {
+			// If we didn't hit this error last time we'll be requeued
+			// implicitly due to the status update. Otherwise we want to retry
+			// after a brief wait, in case this was a transient error.
+			claim.SetConditions(v1alpha1.Creating(), v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: AShortWait}, errors.Wrap(r.client.Status().Update(ctx, claim), "cannot update resource claim status")
+		}
+	}
+
+	if resource.GetBindingPhase() == v1alpha1.BindingPhaseUnbindable {
+		// If this claim was not already binding we'll be requeued due to the
+		// status update. Otherwise there's no need to requeue. We should be
+		// watching both the resource claims and the resources we own, so we'll
+		// be queued if anything changes.
+		claim.SetConditions(Binding(), v1alpha1.ReconcileSuccess())
+		return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, claim), "cannot update resource claim status")
+	}
+
+	if resource.GetBindingPhase() == v1alpha1.BindingPhaseUnbound {
+		if err := r.resource.PropagateConnection(ctx, claim, resource); err != nil {
+			// If we didn't hit this error last time we'll be requeued implicitly
```

Comment:
minor: most of the comments in this function are around potential context across reconciliation attempts (i.e. what could have happened on the last reconciliation before this one), which is great and useful to think through the requeueing scenarios and decisions.  I think though that some more comments about the happy path "what are we doing at each step and why" could serve to make this a bit easier to follow.

Reply:
Could you give some examples of places you think some more commentary would be useful?